exhale/utils.py

```diff
@@ -834,3 +834,74 @@ def fancyError(critical_msg=None, lex="py3tb", singleton_hook=None):
             ))
 
     os._exit(1)
+
+
+def groupsFromBalancedBrackets(string : str, left_bracket : str, right_bracket : str):
+    """Returns a (nested) list of characters with their corresponding children (if any)"""
+    def push(obj, l, depth):
+        while depth:
+            l = l[-1]
+            depth -= 1
+        l.append(obj)
+
+    def pop(l, depth):
+        while depth:
+            l = l[-1]
+            depth -= 1
+        return l.pop()
+
+    groups = []
+    depth = 0
+    try:
+        for idx, char in enumerate(string):
+            if char == left_bracket:
+                prev_char = pop(groups, depth)
+                if prev_char != ' ':  # remove the space before template brackets
+                    push(prev_char, groups, depth)  # put it back if it wasn't a space...
+                push([], groups, depth)
+                depth += 1
+            elif char == right_bracket:
+                depth -= 1
+            else:
+                push(char, groups, depth)
+    except IndexError:
+        raise ValueError('Parentheses mismatch')
+    if depth > 0:
+        raise ValueError('Parentheses mismatch')
+    else:
+        return groups
+
+
+def groupsToNamedGroups(groups, outer_group):
+    """Turns the output of groupsFromBalancedBrackets into a list of strings"""
+    current_string = ""
+    for char in groups:
+        if type(char) is type([]):
+            outer_group.append(current_string)
+            outer_group.append([])
+            groupsToNamedGroups(char, outer_group[-1])
+            current_string = ""
+        else:
+            if char == ',':
+                if current_string != '':
+                    outer_group.append(current_string)
+                    current_string = ''
+            else:
+                current_string += char
+    if current_string != '':
+        outer_group.append(current_string)
+
+
+def templateListToNodeName(template_list):
```

Comment:
This function reconstructs the template name so that we can easily split it on the correct separators later.
Given the output from the above function, it will return:
```py
Image<1920, 1080>
```